app/src/org/commcare/utils/PrivilegesUtility.java

```diff
@@ -0,0 +1,185 @@
+package org.commcare.utils;
+
+import org.commcare.modern.util.Pair;
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+
+import java.util.ArrayList;
+
+/**
+ * Helper methods for privilege data
+ *
+ * Created by ctsims on 8/7/2017.
+ */
+
+public class PrivilegesUtility {
+    private String authorityPublicKeyString;
+
+    public PrivilegesUtility(String authorityPublicKeyString) {
+        this.authorityPublicKeyString = authorityPublicKeyString;
+    }
+
+    public Pair<String, String[]> processPrivilegePayloadForActivatedPrivileges
+            (String privilegePayloadJson) throws PrivilagePayloadException{
+        try {
+            JSONObject obj = new JSONObject(privilegePayloadJson);
+            int version;
+            if(obj.has("version")) {
+                version = obj.getInt("version");
+            } else {
+                if(obj.has("flag")) {
+                    version = 1;
+                } else {
+                    throw new UnrecognizedPayloadVersionException();
+                }
+            }
+
+            switch(version) {
+                case 1:
+                    return processAndValidateV1Payload(obj);
+                case 2:
+                    return processAndValidateV2Payload(obj);
+                default:
+                    throw new UnrecognizedPayloadVersionException();
+            }
+
+        } catch (JSONException e) {
+            throw new MalformedPayloadException(e);
+        }
+    }
+
+    private Pair<String, String[]> processAndValidateV2Payload(JSONObject obj) throws JSONException, PrivilagePayloadException {
```

Comment:
Even though it seems a little repetitive, I think the readability of this method and `procesesAndValidateV1Payload()` would be helped enormously if after calling `processV1Payload()` and `processV2Payload()`, respectively, you created named variables for each of the results, instead of just referencing them by `payload.first`, `payload[0]`, etc.